library/alloc/src/boxed/thin.rs

```diff
@@ -0,0 +1,198 @@
+    /// #![feature(thin_box)]
+    /// use std::boxed::ThinBox;
+    ///
+    /// let five = ThinBox::new(5);
+    /// ```
+    pub fn new(value: T) -> Self {
+        let dyn_ref: &T = &value;
+        let (_, meta) = (dyn_ref as *const T).to_raw_parts();
+        let ptr = WithHeader::new(meta, value).expect("No allocation failure");
+        ThinBox { ptr, _marker: PhantomData }
+    }
+}
+
+#[unstable(feature = "thin_box", issue = "92791")]
+impl<Dyn: ?Sized> ThinBox<Dyn> {
+    /// Moves a type to the heap with it's `Metadata` stored in the heap allocation instead of on
+    /// the stack.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(thin_box)]
+    /// use std::boxed::ThinBox;
+    ///
+    /// let thin_slice = ThinBox::<[i32]>::new_unsize([1, 2, 3, 4]);
+    /// ```
+    pub fn new_unsize<T>(value: T) -> Self
+    where
+        T: Unsize<Dyn>,
+    {
+        let dyn_ref: &Dyn = &value;
+        let (_, meta) = (dyn_ref as *const Dyn).to_raw_parts();
+        let ptr = WithHeader::new(meta, value).expect("No allocation failure");
+        ThinBox { ptr, _marker: PhantomData }
+    }
+}
+
+#[unstable(feature = "thin_box", issue = "92791")]
+impl<T: ?Sized + Debug> Debug for ThinBox<T> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?}", self.deref())
+    }
+}
+
+#[unstable(feature = "thin_box", issue = "92791")]
+impl<T: ?Sized> Deref for ThinBox<T> {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        let value = self.data();
+        let metadata = self.meta();
+        let pointer = ptr::from_raw_parts(value as *const (), metadata);
+        unsafe { &*pointer }
+    }
+}
+
+#[unstable(feature = "thin_box", issue = "92791")]
+impl<T: ?Sized> DerefMut for ThinBox<T> {
+    fn deref_mut(&mut self) -> &mut T {
+        let value = self.data();
+        let metadata = self.meta();
+        let pointer = ptr::from_raw_parts::<T>(value as *const (), metadata) as *mut T;
+        unsafe { &mut *pointer }
+    }
+}
+
+#[unstable(feature = "thin_box", issue = "92791")]
+impl<T: ?Sized> Drop for ThinBox<T> {
+    fn drop(&mut self) {
+        unsafe {
+            let value = self.deref_mut();
+            let value = value as *mut T;
+            self.ptr.drop::<T>(value);
+        }
+    }
+}
+
+#[unstable(feature = "thin_box", issue = "92791")]
+impl<T: ?Sized> ThinBox<T> {
+    fn meta(&self) -> <T as Pointee>::Metadata {
+        //  Safety:
+        //  -   NonNull and valid.
+        unsafe { *self.ptr.header() }
+    }
+
+    fn data(&self) -> *mut u8 {
+        self.ptr.value()
+    }
+}
+
+struct WithHeader<H>(*mut u8, PhantomData<H>);
+
+impl<H> WithHeader<H> {
+    fn new<T>(header: H, value: T) -> Option<WithHeader<H>> {
+        let layout = Self::alloc_layout(mem::size_of::<T>(), mem::align_of::<T>());
+        let aligned_header_size = Self::aligned_header_size(layout.align());
+
+        unsafe {
+            let ptr = NonNull::new(alloc::alloc(layout))?;
```

Comment:
It would be interesting to avoid calling `alloc::alloc` if both `T` and `H` are ZST.

Reply:
I'm not exactly sure how to do that, though that does sound fun. I think I'd like to leave that for a followup PR, though I will add some tests right now to make sure our ZST behavior isn't completely broken.

Reply:
Actually though, what tests can I even run to ensure thinbox is working when the T and H are both ZSTs? My understanding is that would mean that the vtable is empty so it would have to be a trait object for something like `trait Useless {}`. what can I even do with that type?

Reply:
Nevermind, using `alloc::System` does this for us if we add a generic allocator parameter to `ThinBox`, which will be done as a follow-up, so no need to bother about that small optimization now.